test/async_utils_test.py

```diff
@@ -456,3 +458,58 @@ async def foo():
     async for item in callable_to_agen(foo):
         result.append(item)
     assert result == [await foo()]
+
+
+@pytest.mark.asyncio
+async def test_async_map():
+    result = []
+    states = []
+
+    async def foo():
+        states.append("enter")
+        try:
+            yield 1
+            yield 2
+            yield 3
+        finally:
+            states.append("exit")
+
+    async def mapper(x):
+        await asyncio.sleep(0.1)  # Simulate some async work
+        return x * 2
+
+    async for item in async_map(foo(), mapper, concurrency=3):
```

Comment:
We need to extend the test suite for async_map to ensure the results actually come in as soon as they are ready. Turns out the current implementation has a bug (`output_event` is never set!) that cause no items to be returned until the full mapping is complete 😬

Reply:
Should probably extend the test suite for all of these replacement functions tbh, since they are pretty critical to the operations of the client.

Other things that came to mind that we need tests for is what happens if there are exceptions in the mapper function, if there are exceptions in the input iterators, or if the operation itself gets cancelled (as in asyncio.CancelledError).

We'll soon have a pretty useful and stable library we could extract and open source by itself :D